src/ess/reduce/nexus.py

```diff
@@ -0,0 +1,345 @@
+
+
+def _unique_child_group(
+    group: snx.Group, nx_class: Type[snx.NXobject], name: Optional[str]
+) -> snx.Group:
+    if name is not None:
+        child = group[name]
+        if isinstance(child, snx.Field):
+            raise ValueError(
+                f"Expected a NeXus group as item '{name}' but got a field."
+            )
+        if child.nx_class != nx_class:
+            raise ValueError(
+                f"The NeXus group '{name}' was expected to be a "
+                f'{nx_class} but is a {child.nx_class}.'
+            )
+        return child
+
+    children = group[nx_class]
+    if len(children) != 1:
+        raise ValueError(f'Expected exactly one {nx_class} group, got {len(children)}')
+    return next(iter(children.values()))  # type: ignore[return-value]
+
+
+def extract_detector_data(detector: RawDetector) -> RawDetectorData:
+    """Get and return the events or histogram from a detector loaded from NeXus.
+
+    This function looks for a data array in the detector group and returns that.
+
+    Parameters
+    ----------
+    detector:
+        A detector loaded from NeXus.
+
+    Returns
+    -------
+    :
+        A data array containing the events or histogram.
+
+    Raises
+    ------
+    ValueError
+        If there is more than one data array.
+
+    See also
+    --------
+    load_detector:
+        Load a detector from a NeXus file in a format compatible with
+        ``extract_detector_data``.
+    """
+    return RawDetectorData(_extract_events_or_histogram(detector))
+
+
+def extract_monitor_data(monitor: RawMonitor) -> RawMonitorData:
+    """Get and return the events or histogram from a monitor loaded from NeXus.
+
+    This function looks for a data array in the monitor group and returns that.
+
+    Parameters
+    ----------
+    monitor:
+        A monitor loaded from NeXus.
+
+    Returns
+    -------
+    :
+        A data array containing the events or histogram.
+
+    Raises
+    ------
+    ValueError
+        If there is more than one data array.
+
+    See also
+    --------
+    load_monitor:
+        Load a monitor from a NeXus file in a format compatible with
+        ``extract_monitor_data``.
+    """
+    return RawMonitorData(_extract_events_or_histogram(monitor))
+
+
+def _extract_events_or_histogram(dg: sc.DataGroup) -> sc.DataArray:
+    data_arrays = {
+        key: value for key, value in dg.items() if isinstance(value, sc.DataArray)
+    }
+    if len(data_arrays) == 0:
+        raise ValueError(
+            "Raw data loaded from NeXus does not contain events or a histogram. "
+            "Expected to find a data array, "
+            f"but the data only contains {set(dg.keys())}"
+        )
+    if len(data_arrays) > 1:
+        raise ValueError(
+            "Raw data loaded from NeXus contains more than one data array. "
+            "Cannot uniquely identify the event or histogram data. "
+            f"Got items {set(dg.keys())}"
+        )
```

Comment:
This won't do, unfortunately. Facilities have the tendency of storing a histogrammed "preview" of the event data alongside the events. I think we should return the events if found, and the dense data otherwise.